confluent_kafka/consumer.py

```diff
@@ -0,0 +1,162 @@
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from .cimpl import Consumer as _impl
+from warnings import warn
+
+
+class Consumer(_impl):
+    """
+        Create a new Kafka Consumer instance.
+
+        To avoid spontaneous calls from non-Python threads all callbacks will only be served upon
+        calling ```client.poll()``` or ```client.flush()```.
+
+        :param dict conf: Configuration properties.
+            See https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md for more information.
+        :param func key_deserializer(topic, key): Converts message key bytes to object.
+            **note** deserializers are responsible for handling NULL keys
+        :param func value_deserializer(topic, value): Converts message value bytes to object.
+            **note** deserializers are responsible for handling NULL values
+        :param func on_commit(err, [partitions]): Callback used to indicate success or failure
+            of an offset commit.
+        :param func stats_cb(json_str): Callback for statistics emitted every ``statistics.interval.ms``.
+            See https://github.com/edenhill/librdkafka/wiki/Statistics” for more information.
+        :param func throttle_cb(confluent_kafka.ThrottleEvent): Callback for throttled request reporting.
+        :param logging.handlers logger: Forwards logs from the Kafka client to the provided handler instance.
+            Log messages will only be forwarded when ``client.poll()`` or ``producer.flush()`` are called.
+        :raises TypeError: If conf is not a dict.
+    """
+    def __new__(cls, *args, **kwargs):
+        if 'key_deserializer' in kwargs or 'value_deserializer' in kwargs:
+            return super(Consumer, cls).__new__(SerializingConsumer, *args, **kwargs)
+        return super(Consumer, cls).__new__(cls, *args, **kwargs)
+
+
+class SerializingConsumer(Consumer):
+    """
+        SerializingConsumer extends Consumer with configurable key and value deserializer.
+
+        Instances of SerializingConsumer cannot be created directly.
+        To obtain an instance of this class instantiate a Consumer with a key and/or value deserializer.
+
+        Duplicate params have been omitted for brevity. See Consumer for class documentation.
+
+        :raises TypeError: If conf is not a dict.
+        :raises TypeError: If instantiated directly.
+    """
+
+    __slots__ = ["_key_deserializer", "_value_deserializer"]
+
+    def __new__(cls, *args, **kwargs):
+        raise TypeError("SerializingConsumer is a non user-instantiable class")
+
+    @staticmethod
+    def byteSerializer(topic, data):
```

Comment:
best thought of an implementation detail, mark as private.

Reply:
Originally it was set as a parameter default which is why I didn't mark it private. Then I realized bad things will happen if you pass None. Thanks for reminding me to come back to this!